Estimate test split from file source

lvotypko opened the pull request:
Solution for [JENKINS-47206](https://issues.jenkins-ci.org/browse/JENKINS-47206).

jtnord commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +158,41 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
         return true;
     }
 
+    public static Map<String, TestClass>  findTestResultsInDirectory(Run<?,?> build, TaskListener listener, FilePath workspace){
```
method name is confusing as it is not finding test results but test source.

lvotypko replied:
yes you are right, my bad .... I will rename it.

jtnord commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +158,41 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
+            return data;
+        }
+        for(String test : tests){
+            test = test.split("src/test/java/")[1];
```
does this work when running on windows?  (it would appear as though it does not)

jglick replied:
[This](https://ci.jenkins.io/job/Plugins/job/parallel-test-executor-plugin/job/PR-34/1/testReport/junit/org.jenkinsci.plugins.parallel_test_executor/ParallelTestExecutorUnitTest/findTestInJavaProjectDirectory/) suggests not:

```
java.lang.ArrayIndexOutOfBoundsException: 1
	at org.jenkinsci.plugins.parallel_test_executor.ParallelTestExecutor.findTestResultsInDirectory(ParallelTestExecutor.java:181)
	at org.jenkinsci.plugins.parallel_test_executor.ParallelTestExecutor.findTestSplits(ParallelTestExecutor.java:214)
	at org.jenkinsci.plugins.parallel_test_executor.ParallelTestExecutorUnitTest.findTestInJavaProjectDirectory(ParallelTestExecutorUnitTest.java:119)
```

lvotypko replied:
I did not test it on all platforms. I guess that in case of unix system it is alright to test it only for linux. I will run tests for Windows and fix bug. Thanks.

jtnord commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +158,41 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
+                @Override
+                public String[] call() throws Throwable {
+
+                    return Util.createFileSet(new File(baseDir), "**/src/test/java/**/*Test.java").getDirectoryScanner().getIncludedFiles();
```
this does not match the defaults that surefire uses so many tests will be missed.

http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#includes lists

```
<includes>
    <include>**/Test*.java</include>
    <include>**/*Test.java</include>
    <include>**/*Tests.java</include>
    <include>**/*TestCase.java</include>
</includes>
```

lvotypko replied:
I though about it, but I did not want to add there something what should not be there (what is not test), for example this one [1]. My first though was to have at least something in the first run (even if it means not to be effective and run some test in every split) and consult the more accurate version in different pullrequest. But if you agree to have it even if it means that we can catch something more then only tests (it is not like the maven cares about it in the end or it can affect a result or performance), I totally agree with this idea.

[1] https://github.com/jenkinsci/junit-plugin/blob/master/src/test/java/hudson/tasks/test/helper/TestResultsPage.java

jtnord replied:
well if you are using exclusions you would run the same tests multiple times for the tests you do not find.
for inclusions you would fail to run some tests.

both cases are bad and if you found too much it should be much less of an issue than not finding enough, unless you happened to have a bucket with some files but actually contained no tests, which would cause a build failure.

jglick commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +158,41 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
     }
 
+    public static Map<String, TestClass>  findTestResultsInDirectory(Run<?,?> build, TaskListener listener, FilePath workspace){
+        if(workspace==null){
```
Mark it `@CheckForNull`.

jglick commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +158,41 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
     static List<InclusionExclusionPattern> findTestSplits(Parallelism parallelism, Run<?,?> build, TaskListener listener,
                                                           boolean generateInclusions,
-                                                          @CheckForNull final String stageName) {
+                                                          @CheckForNull final String stageName, FilePath workspace) {
```
Mark nullability on `workspace`.

jglick commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java:
```diff
@@ -75,13 +76,14 @@ public DescriptorImpl() {
         @StepContextParameter private transient TaskListener listener;
 
         @Override protected List<?> run() throws Exception {
+            FilePath path = this.getContext().get(FilePath.class);
```
Note that this will be null if not inside `node`. (The usual case for existing scripts not prepared to take advantage of this feature.)

You may also want to update [this](https://github.com/jenkinsci/parallel-test-executor-plugin/blob/a5910785979c4b40e02e7fc4ebe8ecc69fdc5c54/demo/lib/vars/testInParallel.groovy#L2) to read something like

```groovy
def splits
node {
  prepare()
  splits = splitTests parallelism: parallelism, generateInclusions: true
}
```

and adjust the [documentation](https://github.com/jenkinsci/parallel-test-executor-plugin/blob/a5910785979c4b40e02e7fc4ebe8ecc69fdc5c54/demo/README.md#demo-contents) to note that even the first build (or the first build of a new branch) will have some reasonable default split.

lvotypko replied:
Of course. I will add an example how to use it - it should not be obvious for user that this feature does not have sense without node and project source code.

lvotypko replied:
Jesee, I am not sure what you meant by this "Was expecting something in ParallelTestExecutorTest demonstrating the effectiveness of the new feature in the context of an actual Pipeline build". Do you wish a test which compares two pipeline jobs execution (one without new feature and one with new feature) in the matter of time?

jglick commented:
> I am not sure what you meant by this "Was expecting something in ParallelTestExecutorTest demonstrating the effectiveness of the new feature in the context of an actual Pipeline build". Do you wish a test which compares two pipeline jobs execution (one without new feature and one with new feature) in the matter of time?

Sorry, no, I meant something pretty simple: existing test cases [like this](https://github.com/jenkinsci/parallel-test-executor-plugin/blob/a5910785979c4b40e02e7fc4ebe8ecc69fdc5c54/src/test/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutorTest.java#L52-L58) are resigned to the fact that there is only a single bucket returned from the first build and the plugin only works well on the second build; in this PR, you could write a new test which runs `splitTests` inside a `node` full of sources (most easily prepopulated using `Node.getTopLevelWorkspaceFor(Item)`) and assert that even the first build calculates a reasonable and useful split.

This PR looks like a useful and very oft-requested feature, so do not let my lack of proper review hold you back from releasing if you think it is ready.

lvotypko commented:
@raul-arabaolaza , I tested it on only one windows so you can be right. But I am not sure why you want to make this feature optional this way, because it is already optional according to my opinion. You have to use it inside node with pre-provided workspace with file sources you desire to be used. I think that it makes it optional enough (or better intentional) with comparing of example settings in demo. And I would like to have support for windows too, so some report from those who use it and it fails would be very useful -s would not agree to do some if and give up in case of windows OS. But I can make it this option more visible for user - user has to directly ask for it by some parameter.

@jglick, it is not problem, I can easily provided such test by reusing the already written test (just add some mess to the workspace and check what exactly is returned as result).

lvotypko commented:
@raul-arabaolaza, great, I will do it in next request. I added required test and I am really glad because I found out that I forgot that filter is in the workspace not in the test folder (so I can not use exactly the maven pattern, but I have to add the path to the test folder in maven project) - so I changed pattern - thank you Jesse!

jglick commented:
Again do not let my current lack of time to review block you from releasing something useful.

lvotypko commented:
@jglick Thank you Jesse, I take it as an approval for merge.
@raul-arabaolaza  can you do it? I do not have rights for it.

raul-arabaolaza commented:
@lvotypko Sorry we are in the same situation, as per the commit history we need to wait for @abayer or @jglick for merge

jtnord commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +161,60 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
+        testFilesExpression.add("**/src/test/java/**/*Tests.java");
+        testFilesExpression.add("**/src/test/java/**/*TestCase.java");
+        try {
+            separator = workspace.act(new MasterToSlaveCallable<String, Throwable>() {
```
you could have just used a regular expression and avoid a callable.  `[\/]`  (would obviously need the correct escaping applied)

lvotypko replied:
I would let it be as it is. There is a big probability that this part will be changed when it was used on Windows according to reported issues.

jtnord commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java:
```diff
@@ -41,6 +44,15 @@ public void setGenerateInclusions(boolean generateInclusions) {
+        this.estimateTestsFromFiles = estimateTestsFromFiles;
+    }
+
+    public boolean getEstimateTestsFromFiles(){
```
should be `isEstimateTestsFromFiles`

lvotypko replied:
Agree, better to change now.

agentgonzo commented:
This is a really welcome addition 👍

lvotypko commented on src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java:
```diff
@@ -155,14 +161,60 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
+            String path = StringUtils.join(new String[]{"src", "test", "java"}, separator);
+            test = test.split(path + separator)[1];
+            //remove suffix of file
+            test = test.substring(0,test.length()-5);
```
Agree, it is better readable then splitting.
